Implementa Cheque.pasamanos y otras mejoras en cheques

fauno opened the pull request.

fauno committed:
Las chequeras son Cajas

fauno committed:
Los cheques tienen un destino

fauno committed:
Renombrar recibo_id a origen_id #55

fauno committed:
Mantenemos el nombre recibo para no tener que cambiar las vistas

fauno committed:
Mover el test de deposito de cheques a cheques

fauno committed:
Los cheques de terceros van en una caja chequera

fauno committed:
Las extracciones y los depositos ahora devuelven recibos

fauno committed:
Los cheques tienen dos recibos y una sola caja #55

Al depositar un cheque se cambia de caja y se asocia a un recibo interno
incompleto.

fauno committed:
La transacción de deposito no tiraba bien el rollback

fauno committed:
Pasamanos! Me faltan algunos tests más piolas, pero ya no puedo seguir #55

fauno committed:
Comprobar que se haya hecho la transferencia

m4tu5 commented on app/models/cheque.rb:
```diff
@@ -57,45 +63,100 @@ def pagado?
-    return nil unless self.propio?
     # Los cheques pagados no se pueden pagar dos veces!
     return nil if self.pagado?
+    # solo los cheques de terceros que se pasaron de manos se pueden
```
tambien los cheques propios se pueden pagar. 

los cheques que pasaron de manos ya no son nuestro problema. los paga quien los emitio, no nosotros, aunque entiendo que este metodo se use para movimentar la caja asociada.

fauno replied:
el comentario se refiere al test no al modelo

m4tu5 commented on app/models/cheque.rb:
```diff
@@ -57,45 +63,100 @@ def pagado?
+    # solo los cheques de terceros se pasan de manos
+    return nil unless self.terceros?
+    # no tienen que estar vencidos
+    return nil if self.vencido?
```
no hay problemas en que un cheque de un tercero este vencido para pasarlo de manos ... es como un cheque al día. este control no es necesario (y podria generar algun conflicto a futuro).

lo que no se puede hacer es depositar un cheque que lleve mas de 30 dias vencido, con lo que nos puede interesar identificar los cheques que lleven vencidos 15 o 20 dias para depositarlos a tiempo y no perderlos ... pero esto puede hacerse mas adelante.

m4tu5 commented on app/models/cheque.rb:
```diff
@@ -57,45 +63,100 @@ def pagado?
+    Cheque.transaction do
+      self.destino = recibo_destino
+      self.estado = 'pasamanos'
+    end
```
el cheque de 3ros, una vez pasado de manos deja de ser nuestro problema y sale de la chequera, con lo que creo que aca va un

self.pagar

ya que el pasamanos implica el pago inmediato de la deuda y su consecuente salida de dinero de la chequera

m4tu5 commented on test/models/cheques_test.rb:
```diff
@@ -31,8 +31,9 @@ class ChequeTest < ActiveSupport::TestCase
+    assert_not cheque.depositar(caja)
   end
 
   test "los cheques de terceros no se pagan" do
```
esto ya no es asi. ahora los cheques de 3ros, si son pasamanos, si se pagan

m4tu5 commented on test/models/cheques_test.rb:
```diff
@@ -51,4 +52,57 @@ class ChequeTest < ActiveSupport::TestCase
     assert_equal monto * -1, cheque.recibo.movimientos.last.monto
```
imagino que este test funciona porque el movimiento de descuento es el ultimo que se crea, pero el test iria mejor si testeasemos que existe un movimiento con monto negativo por valor del cheque

tal vez algo asi?
assert cheque.recibo.movimiento.where(monto: ( monto \* -1)).any?
o
assert @cheque.movimientos.collect(&:monto).include?(Money.new(monto \* -1))

m4tu5 commented on test/models/cheques_test.rb:
```diff
@@ -51,4 +52,57 @@ class ChequeTest < ActiveSupport::TestCase
+
+    # asegurar que haya dinero disponible en la caja
+    assert caja.depositar(cheque.monto)
+    assert cheque.monto, caja.total
```
assert_equal?

m4tu5 commented on test/models/cheques_test.rb:
```diff
@@ -51,4 +52,57 @@ class ChequeTest < ActiveSupport::TestCase
+    assert_equal 2, recibo_interno.movimientos.count, cheque.inspect
+    assert_equal cheque.monto * -1, recibo_interno.movimientos.first.monto
+    assert_equal cheque.monto, recibo_interno.movimientos.last.monto
+
```
lo mismo que arriba, no seria mejor que los test no dependan del orden en que se guardan los movimientos?

mauriciopasquier replied:
+1

fauno replied:
Mauricio Pasquier Juan notifications@github.com writes:

> +1

no entendi a que +1neas, al test?

## 

:>

fauno replied:
ah el mail me llegó con el código pero sin el comentario de matus

m4tu5 commented on test/models/cheques_test.rb:
```diff
@@ -51,4 +52,57 @@ class ChequeTest < ActiveSupport::TestCase
+
+  end
+
+  test 'pasar un cheque de manos y despues cobrarlo' do
```
test ' ... y despues PAGARLO' do

m4tu5 commented on test/models/cheques_test.rb:
```diff
@@ -51,4 +52,57 @@ class ChequeTest < ActiveSupport::TestCase
+    cheque = create :cheque, situacion: 'terceros', caja: chequera
+    recibo_de_pago = create :recibo, situacion: 'pago'
+
+    assert chequera.depositar(cheque.monto)
```
no deberíamos hacer que el modelo haga el deposito automaticamente al crear el cheque (en 'chequera' si es de 3ros, y en '@cheque.caja' si es propio)?

tal vez incorporarlo como parte de 'emitir_cheque'

fauno replied:
m4tu5 notifications@github.com writes:

> no deberíamos hacer que el modelo haga el deposito automaticamente al
> crear el cheque (en 'chequera' si es de 3ros, y en '@cheque.caja' si
> es propio)?

es lo que estabamos discutiendo ayer con @mauriciopasquier:

19:37:10    mpj   cuando hablás de generar depósitos en 'chequeras'
estás diciendo crear movimientos?
19:37:42  fauno   claro
19:37:50    mpj   ok
19:38:27          sí, creás un movimiento cuando creás el cheque, pero no
lo crea el cheque, se crea a través del recibo  
19:38:27  fauno   pense que estaba claro que caja.depositar es generar
un movimiento che
19:38:35    mpj   con el mecanismo de cobro de facturas/recibos
19:38:57  fauno   o sea que cada vez que creas un cheque de terceros
tambien tenes que crear el deposito en chequera a mano?
19:39:16    mpj   ?
19:39:22          "a mano"?
19:39:47  fauno   no se como estas pensando la interfaz de crear
movimientos en realidad
                  digo si vas a hacer Cheque.create(situacion: terceros)
lo tenes que acompañar con un Caja.depositar
19:41:29          en lugar de que crear un cheque de terceros se
encargue solo de depositarse en la chequera
19:42:01    mpj   lo veremos cuando hagamos la interfaz
19:42:06          hay muchas formas de implementarlo 

## 

:>

fauno committed:
assert_equals

fauno committed:
no importa si los cheques de terceros están vencidos

"no hay problemas en que un cheque de un tercero este vencido para
pasarlo de manos ... es como un cheque al día. este control no es
necesario (y podria generar algun conflicto a futuro)."

fauno committed:
el cheque pasado de manos se paga inmediatamente

"el cheque de 3ros, una vez pasado de manos deja de ser nuestro problema
y sale de la chequera, con lo que creo que aca va un

self.pagar

ya que el pasamanos implica el pago inmediato de la deuda y su
consecuente salida de dinero de la chequera "

fauno committed:
a veces los cheques de terceros se pagan

fauno committed:
no importa el orden de los movimientos

fauno committed:
sentido
